### PR TITLE
[geometry] Added different bounding boxes.

### PR DESCRIPTION
**Pull request by ygorshenin**

### CODE REVIEW

File: geometry/segment2d.cpp [hunk 1]
```diff
@@ -2,10 +2,16 @@
 
 #include "geometry/robust_orientation.hpp"
 
-#include "std/algorithm.hpp"
+#include <algorithm>
```
Comment (tatiana-kondakova) [inline geometry/segment2d.cpp]: Empty line between `include` and `using namespace` please

File: geometry/CMakeLists.txt [hunk 2]
```diff
@@ -8,18 +8,25 @@ set(
   angles.hpp
   any_rect2d.hpp
   avg_vector.hpp
+  bbox.cpp
```
Comment (mpimenov) [inline geometry/CMakeLists.txt]: These names don't seem descriptive enough. While "bbox" may pass as a common term, "cbox" and "dbox" are surely too narrow. Please, either rename the files or create a subdirectory for them and an explanatory comment.

File: geometry/bbox.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include "geometry/point2d.hpp"
+
+#include <limits>
+#include <vector>
+
+namespace m2
+{
+class BBox
```
Comment (mpimenov) [inline geometry/bbox.hpp]: How is this class different from m2::RectD?

File: geometry/cbox.hpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "geometry/point2d.hpp"
+
+#include <vector>
+
+namespace m2
+{
+class CBox
```
Comment (mpimenov) [inline geometry/cbox.hpp]: Please provide a short description of this class.

File: geometry/cbox.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+// Checks whether (p1 - p) x (p2 - p) >= 0.
+bool IsCCW(PointD const & p1, PointD const & p2, PointD const & p)
+{
+  return robust::OrientedS(p1, p2, p) > -kEps;
+}
+
+PointD Ort(PointD const & p) { return PointD(-p.y, p.x); }
+
+template <typename Fn>
+void ForEachRect(ConvexHull const & hull, Fn && fn)
```
Comment (mpimenov) [inline geometry/cbox.cpp]: Please provide a short description of this function.

File: geometry/dbox.hpp [hunk 6]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include "geometry/bbox.hpp"
+#include "geometry/point2d.hpp"
+
+namespace m2
+{
+class DBox
```
Comment (mpimenov) [inline geometry/dbox.hpp]: Please provide a short description of this class.

File: geometry/line2d.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+#include <cmath>
+#include <sstream>
+
+using namespace std;
+
+namespace m2
+{
+namespace
+{
```
Comment (tatiana-kondakova) [inline geometry/line2d.cpp]: I think it'll be better to move anonymous namespace out of namespace m2

File: geometry/geometry_tests/line2d_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+  {
+    Line2D const lhs(Segment2D(PointD(0, 0), PointD(1, 1)));
+    Line2D const rhs(Segment2D(PointD(-10, -10), PointD(-100, -100)));
+    TEST_EQUAL(Intersect(lhs, rhs).m_type, Type::Infinite, ());
+  }
+
+  {
+    Line2D const lhs(Segment2D(PointD(0, 0), PointD(1, 1)));
+    Line2D const rhs(Segment2D(PointD(-10, -10), PointD(-100, -100)));
```
Comment (tatiana-kondakova) [inline geometry/geometry_tests/line2d_tests.cpp]: Same test twice. Here and 6 lines before.

File: geometry/line2d.cpp (hunk 7 above)
Comment (ygorshenin) [reply geometry/line2d.cpp]: Why? What's the problem?

File: geometry/geometry_tests/line2d_tests.cpp (hunk 8 above)
Comment (ygorshenin) [reply geometry/geometry_tests/line2d_tests.cpp]: Done

File: geometry/segment2d.cpp (hunk 1 above)
Comment (ygorshenin) [reply geometry/segment2d.cpp]: Done

File: geometry/bbox.hpp (hunk 3 above)
Comment (ygorshenin) [reply geometry/bbox.hpp]: Bounding box is a bounding box around some set of points, RectD is a rectangle, not related to points on the plane at all.

File: geometry/line2d.cpp (hunk 7 above)
Comment (tatiana-kondakova) [reply geometry/line2d.cpp]: Если мы так ещё где-то делаем и вообще у нас так принято, то ок (пока в коде проекта не встречала). Технически понятно что эффект тот же. Up to you.

File: geometry/CMakeLists.txt (hunk 2 above)
Comment (ygorshenin) [reply geometry/CMakeLists.txt]: Please, suggest a name for cbox. I'm going to rename bbox to a bounding box and dbox to a diamond box.

File: geometry/line2d.cpp (hunk 7 above)
Comment (ygorshenin) [reply geometry/line2d.cpp]: Я в любом случае не понимаю какую проблему создает анонимный namespace в source файле внутри другого namespace. Я понимаю какую проблему это решает - можно избавиться от назойливых ```m2::``` префиксов.

File: geometry/bbox.cpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#include "geometry/bbox.hpp"
+
+#include <algorithm>
+using namespace std;
+
+namespace m2
+{
+BBox::BBox(std::vector<PointD> const & points)
```
Comment (tatiana-kondakova) [inline geometry/bbox.cpp]: we `using namespace std` here

File: geometry/bbox.cpp [hunk 10]
```diff
@@ -0,0 +1,26 @@
+#include "geometry/bbox.hpp"
+
+#include <algorithm>
+using namespace std;
```
Comment (tatiana-kondakova) [inline geometry/bbox.cpp]: Empty line between `include` and `using namespace` please

File: geometry/cbox.hpp [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+{
+// When a size of a convex hull over a set of |points| is less than 4,
+// stores convex hull explicitly, otherwise stores smallest rectangle
+// containing the hull.
+class CBox
+{
+public:
+  CBox(std::vector<PointD> const & points);
+
+  std::vector<PointD> Points() const { return m_points; }
```
Comment (tatiana-kondakova) [inline geometry/cbox.hpp]: is it possible to return `std::vector<PointD> const &` here?

Comment (ygorshenin) [reply geometry/cbox.hpp]: Sure, done.

File: geometry/bbox.cpp (hunk 10 above)
Comment (ygorshenin) [reply geometry/bbox.cpp]: Done

File: geometry/bbox.cpp (hunk 9 above)
Comment (ygorshenin) [reply geometry/bbox.cpp]: Done

File: geometry/dbox.hpp (hunk 6 above)
Comment (ygorshenin) [reply geometry/dbox.hpp]: Done

File: geometry/cbox.cpp (hunk 5 above)
Comment (ygorshenin) [reply geometry/cbox.cpp]: Done

File: geometry/cbox.hpp (hunk 4 above)
Comment (ygorshenin) [reply geometry/cbox.hpp]: Done

**Comment by ygorshenin** [comment]
@mpimenov @tatiana-kondakova PTAL

File: geometry/calipers_box.cpp [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    if (area < bestArea)
+    {
+      bestArea = area;
+      bestPoints = points;
+    }
+  });
+
+  if (bestPoints.empty())
+  {
+    BoundingBox bbox(points);
```
Comment (tatiana-kondakova) [inline geometry/calipers_box.cpp]: How can we get here?

File: geometry/calipers_box.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+// Checks whether (p1 - p) x (p2 - p) >= 0.
+bool IsCCW(PointD const & p1, PointD const & p2, PointD const & p)
+{
+  return robust::OrientedS(p1, p2, p) > -kEps;
+}
+
+PointD Ort(PointD const & p) { return PointD(-p.y, p.x); }
+
+// For each facet of the |hull| calls |fn| with the smallest rectangle
+// containing the hull and with one side colliear to the facet.
```
Comment (tatiana-kondakova) [inline geometry/calipers_box.cpp]: s/colliear/collinear

Comment (ygorshenin) [reply geometry/calipers_box.cpp]: Done

File: geometry/calipers_box.cpp (hunk 12 above)
Comment (ygorshenin) [reply geometry/calipers_box.cpp]: We're compiling our project with "fast-math" flag, therefore all assumptions about IEEE754 behavior may be ruined. So, this is just for safety that we leaving object in the valid state.

Comment (ygorshenin) [reply geometry/calipers_box.cpp]: Also, this may be the case when the precision of input data exceeds 1e-12.

File: geometry/calipers_box.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+    auto const oab = Ort(ab);
+    array<Line2D, 4> const lines = {{Line2D(hull.PointAt(i), ab), Line2D(hull.PointAt(j), oab),
+                                     Line2D(hull.PointAt(k), ab), Line2D(hull.PointAt(l), oab)}};
+    vector<PointD> corners;
+    for (size_t u = 0; u < lines.size(); ++u)
+    {
+      for (size_t v = u + 1; v < lines.size(); ++v)
+      {
+        auto result = LineIntersector::Intersect(lines[u], lines[v], kEps);
```
Comment (tatiana-kondakova) [inline geometry/calipers_box.cpp]: а почему тут пересечения ищутся двумя вложенными циклами? почему не один? `lines[0]` с `lines[1]`, `lines[1]` c `lines[2]`...? Lines заполнен буквально в предыдущей строчке(там же у lines[0] и lines[2] задано параллельное направление, а у lines[1] и lines[3] перпендикулярное им) и перестраховываться таким образом кажется излишне

Comment (ygorshenin) [reply geometry/calipers_box.cpp]: Up to you, имхо это экономия на спичках.

Comment (ygorshenin) [reply geometry/calipers_box.cpp]: Removed.

Comment (tatiana-kondakova) [reply geometry/calipers_box.cpp]: лично мне кажется, что один цикл понятнее (по крайней мере для читающего код): нашли точки, лежащие на описанном прямоугольнике, построили через них линии, уложили эти линии в правильном порядке в array, пересекли соседние

File: geometry/calipers_box.cpp [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+  if (n == 0)
+    return false;
+
+  if (n == 1)
+    return AlmostEqualAbs(m_points[0], p, kEps);
+
+  if (n == 2)
+    return IsPointOnSegmentEps(p, m_points[0], m_points[1], kEps);
+
+  for (size_t i = 0; i < n; ++i) {
```
Comment (mpimenov) [inline geometry/calipers_box.cpp]: {
}

File: geometry/calipers_box.hpp [hunk 16]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "geometry/point2d.hpp"
+
+#include <vector>
+
+namespace m2
+{
+// When a size of a convex hull over a set of |points| is less than 4,
```
Comment (mpimenov) [inline geometry/calipers_box.hpp]: s/a size/the size/
s/a convex hull/the convex hull/

s/a set/the set/
or
s/a set of //

File: geometry/calipers_box.hpp [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+// stores convex hull explicitly, otherwise stores smallest rectangle
+// containing the hull.
```
Comment (mpimenov) [inline geometry/calipers_box.hpp]: Please add a note that the rectangle does not necessarily has its sides parallel to the coordinate axes.

File: geometry/diamond_box.hpp [hunk 18]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "geometry/bounding_box.hpp"
+#include "geometry/point2d.hpp"
+
+namespace m2
+{
+// Bounding box for a set of points on the plane, rotated by 45
+// degrees.
```
Comment (mpimenov) [inline geometry/diamond_box.hpp]: Not really: if you take the bounding box and rotate it, the resulting rectangle may differ from the DiamondBox. Besides, there may be bounding boxes different from "the" bounding box.

The smallest rectangle containing |points| whose sides are parallel to the axes rotated by 45 degrees.
or
The smallest rectangle containing |points| whose sides form 45-degree angles with coordinate axes.

File: geometry/calipers_box.hpp (hunk 17 above)
Comment (mpimenov) [reply geometry/calipers_box.hpp]: This box may need a mention that the rectangle stores its points in the counterclockwise order.

File: geometry/calipers_box.cpp [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+double const kInf = numeric_limits<double>::infinity();
+double const kEps = 1e-12;
+
+// Checks whether (p1 - p) x (p2 - p) >= 0.
+bool IsCCW(PointD const & p1, PointD const & p2, PointD const & p)
+{
+  return robust::OrientedS(p1, p2, p) > -kEps;
+}
+
+PointD Ort(PointD const & p) { return PointD(-p.y, p.x); }
```
[the remark this answers is not in the document]
Comment (ygorshenin) [reply geometry/calipers_box.cpp]: From my point of view, current name is good enough - short, local, descriptive.

File: geometry/calipers_box.cpp (hunk 15 above)
Comment (ygorshenin) [reply geometry/calipers_box.cpp]: Wow, how I missed it?... Done

File: geometry/calipers_box.hpp (hunk 16 above)
Comment (ygorshenin) [reply geometry/calipers_box.hpp]: Done

File: geometry/calipers_box.hpp (hunk 17 above)
Comment (ygorshenin) [reply geometry/calipers_box.hpp]: Done

File: geometry/diamond_box.hpp (hunk 18 above)
Comment (ygorshenin) [reply geometry/diamond_box.hpp]: From my point of view in the common sense the term "Bounding Box" means exactly what BoundingBox class does, i.e. the smallest rectangle containing all points with sides parallel to the axes, therefore my comment does not contradict this definition.

File: geometry/calipers_box.cpp (hunk 14 above)
Comment (ygorshenin) [reply geometry/calipers_box.cpp]: Done, done :)

**Comment by ygorshenin** [comment]
@mpimenov PTAL

File: geometry/diamond_box.hpp (hunk 18 above)
Comment (mpimenov) [reply geometry/diamond_box.hpp]: May be; yet the comment sounds ambiguous to me.
The code is simple enough, so it is probably not a problem. Up to you.

**Pull request merged**